pkg/driver/controller_test.go

```diff
@@ -1719,3 +1719,94 @@ func TestControllerUnpublishVolume(t *testing.T) {
 		t.Run(tc.name, tc.testFunc)
 	}
 }
+
+func TestExpandVolume(t *testing.T) {
+	testCases := []struct {
+		name     string
+		req      *csi.ControllerExpandVolumeRequest
+		newSize  int64
+		expResp  *csi.ControllerExpandVolumeResponse
+		expError bool
+	}{
+		{
+			name: "success normal",
+			req: &csi.ControllerExpandVolumeRequest{
+				VolumeId: "vol-test",
+				CapacityRange: &csi.CapacityRange{
+					RequiredBytes: 5 * util.GiB,
+				},
+			},
+			expResp: &csi.ControllerExpandVolumeResponse{
+				CapacityBytes: 5 * util.GiB,
+			},
+		},
+		{
+			name:     "fail empty request",
+			req:      &csi.ControllerExpandVolumeRequest{},
+			expError: true,
+		},
+		{
+			name: "fail exceeds limit after round up",
+			req: &csi.ControllerExpandVolumeRequest{
+				VolumeId: "vol-test",
+				CapacityRange: &csi.CapacityRange{
+					RequiredBytes: 5*util.GiB + 1, // should round up to 6 GiB
+					LimitBytes:    5 * util.GiB,
+				},
+			},
+			expError: true,
+		},
+		{
+			name: "fail negative size from ResizeDisk",
+			req: &csi.ControllerExpandVolumeRequest{
+				VolumeId: "vol-test",
+				CapacityRange: &csi.CapacityRange{
+					RequiredBytes: 5 * util.GiB,
+				},
+			},
+			newSize:  -1,
+			expError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
```

Comment:
We recently refactor the test to create a `testFunc` for each test cases, so that the `testFunc` body is cleaner and easier to maintain when isolated between different test cases. Could you do the same? See `TestControllerUnpublishVolume` as an example.

Reply:
hm, I agree that having a main test function might make things hard to read sometimes,  but the `testFunc` from other tests has lots of duplicated boilerplate code.

It seems like just adding the `testFunc` added to this file around 200 lines of boilerplate code (d6a4017a). IMO that pattern is more appropriate when changes take 1 or 2 lines.

I can change this test if you feel sure about it, though.

Reply:
Don't want to block this PR. Ok with live with shared test function.